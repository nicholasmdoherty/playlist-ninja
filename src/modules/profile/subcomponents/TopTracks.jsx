import React, { Component } from "react";
import { Image, Col, Row } from "react-bootstrap";
import { connect } from "react-redux";
import { deepCamelCaseKeys } from "../../../common/constants";
import Spacer from "../../../common/components/Spacer";

class TopTracks extends Component {
  constructor(props) {
    super(props);

    this.state = {
      topTracks: null,
      error: null
    };
  }

  componentDidMount() {
    this.props.api.spotifyApi.getMyTopTracks({ limit: 12 }).then(
      data => {
        this.setState({ topTracks: data.body });
      },
      error => {
        this.setState({ error: deepCamelCaseKeys(error) });
      }
    );

    this.forceUpdate();
  }

  render() {
    return (
      <Row className="p-3">
        <Col xs={12} className="text-center">
          <h3 className="display-5"> Top Tracks </h3>
          <Spacer percentage={4} />
        </Col>

        {this.state.topTracks &&
          this.state.topTracks.items.map(topTrack => {
            return (
              <Col xs={6} sm={4} lg={2} className="text-center">
                <div className="p-2 word-wrap">
                  <Image
                    src={topTrack.album.images[0].url}
                    fluid
                    className="drop-shadow mb-3"
                  />
                  <p className="lead">{topTrack.name}</p>
                </div>
              </Col>
            );
          })}
      </Row>
    );
  }
}

let mapStateToProps = state => {
  return {
    api: state.api
  };
};

export default connect(mapStateToProps)(TopTracks);
