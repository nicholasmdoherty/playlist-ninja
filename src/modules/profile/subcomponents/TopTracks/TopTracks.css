.top-tracks-container {
  display: flex;
  flex-direction: column;
  gap: 10px
}

.top-tracks-header {
  font-family: 'Roc Grotesk';
  font-style: normal;
  font-weight: 700;
  font-size: 24px;
  line-height: 35px;
}

.top-tracks-card {
  padding: 15px;
  display: flex;
  flex-direction: column;
  gap: 15px;
}

.top-tracks-button-row {
  display: flex;
  gap: 10px;
}

.top-tracks-button {
  flex-grow: 1;
  background: #353535;
  box-shadow: 0px 2px 5px rgba(0, 0, 0, 0.25);
  padding: 10px 15px;
  align-items: center;
  justify-content: center;
  border-radius: 15px;
  text-align: center;
  font-size: 14px;
}

.top-tracks-button:not(.active):hover {
  cursor: pointer;
  background: linear-gradient(rgba(0,0,0, .1), rgba(0,0,0,.1)) #353535;
  text-decoration: underline;
}

.top-tracks-button.active {
  font-weight: 800;
  background: #09C053; 
}

.top-track-row {
  display: flex;
  flex-direction: row;
  gap: 10px;
  align-items: center;
}

.top-track-name {
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;  
  overflow: hidden;
  line-clamp: 2;
  text-overflow: ellipsis;
  line-height: 18px;
  font-size: 16px;
  max-width: 100%;
}

.top-track-image {
  border-radius: 5px;
}