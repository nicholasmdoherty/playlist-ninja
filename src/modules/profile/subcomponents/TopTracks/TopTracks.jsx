import React, { useEffect, useState } from "react";
import { connect } from "react-redux";

import './TopTracks.css'

const TopTracks = ({ api }) => {

  const [topTracks4Weeks, setTopTracks4Weeks] = useState([])
  const [topTracks6Months, setTopTracks6Months] = useState([])
  const [topTracksAllTime, setTopTracksAllTime] = useState([])
  const [activeView, setActiveView] = useState('short_term')
  const [error, setError] = useState(null)

  useEffect(() => {
    api.getMyTopTracks({ limit: 5, time_range: 'short_term' }).then(
      data => {
        setTopTracks4Weeks(data.body.items)
      },
      error => {
        setError(error)
      }
    );

    api.getMyTopTracks({ limit: 5, time_range: 'medium_term' }).then(
      data => {
        setTopTracks6Months(data.body.items)
      },
      error => {
        setError(error)
      }
    );

    api.getMyTopTracks({ limit: 5, time_range: 'long_term' }).then(
      data => {
        setTopTracksAllTime(data.body.items)
      },
      error => {
        setError(error)
      }
    );
  }, [api])

  const renderTopTracks = () => {

    const renderTrack = (track) => {
      return (
        <div className="top-track-row">
          <img className="top-track-image" width={40} height={40} src={track.album.images.length > 0 ? track.album.images[0].url : ''} />
          <span className="top-track-name">{track.name}</span>
        </div>
      )
    }

    if (error) {
      return 'Error loading top tracks.'
    } else if (!topTracks4Weeks || !topTracks6Months || !topTracksAllTime) {
      return 'Loading...'
    } else if (activeView === 'short_term') {
      return (
        <React.Fragment>
          {topTracks4Weeks.map((track) => {
            return renderTrack(track)
          })}
        </React.Fragment>
      )
    } else if (activeView === 'medium_term') {
      return (
        <React.Fragment>
          {topTracks6Months.map((track) => {
            return renderTrack(track)
          })}
        </React.Fragment>
      )
    } else if (activeView === 'long_term') {
      return (
        <React.Fragment>
          {topTracksAllTime.map((track) => {
            return renderTrack(track)
          })}
        </React.Fragment>
      )
    }
  }

  return (
    <div className='top-tracks-container'>
      <span className="top-tracks-header">Top Tracks</span>
      <div className='top-tracks-button-row'>
        <div
          className={`top-tracks-button ${activeView === 'short_term' ? 'active' : ''}`}
          onClick={() => { setActiveView('short_term') }}
        >
          4 Weeks
        </div>
        <div
          className={`top-tracks-button ${activeView === 'medium_term' ? 'active' : ''}`}
          onClick={() => { setActiveView('medium_term') }}
        >
          6 Months
        </div>
        <div
          className={`top-tracks-button ${activeView === 'long_term' ? 'active' : ''}`}
          onClick={() => { setActiveView('long_term') }}
        >
          All Time
        </div>
      </div>
      <div className='card-style top-tracks-card'>
        {renderTopTracks()}
      </div>
    </div>
  )
}

let mapStateToProps = state => {
  return {
    api: state.api.spotifyApi
  };
};

export default connect(mapStateToProps)(TopTracks);
