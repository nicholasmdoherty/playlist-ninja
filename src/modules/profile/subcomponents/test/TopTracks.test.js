// Package imports
import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";

// Project Imports
import TopTracks from "../TopTracks";
import configureStore from "../../../../store";

describe("TopTracks.jsx", () => {
  /**
   * Tests for the component with no top tracks.
   *  - New spotify account, no data
   */
  describe("with empty track array", () => {
    it("matches the snapshot", () => {
      const component = renderer.create(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopTracks: () => {
                  return Promise.resolve({ body: { items: [] } });
                }
              }
            }
          })}
        >
          <TopTracks />
        </Provider>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders without crashing", () => {
      const div = document.createElement("div");
      ReactDOM.render(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopTracks: () => {
                  return Promise.resolve({ body: { items: [] } });
                }
              }
            }
          })}
        >
          <TopTracks />
        </Provider>,
        div
      );
      ReactDOM.unmountComponentAtNode(div);
    });
  });

  /**
   *  Tests for the component with top track data
   */
  describe("recieved top track data", () => {
    it("matches the snapshot", () => {
      const component = renderer.create(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopTracks: () => {
                  return Promise.reject({
                    body: {
                      items: [
                        { images: null, name: "No Image" },
                        { images: [{ url: "fakeUrl" }], name: "Image" }
                      ]
                    }
                  });
                }
              }
            }
          })}
        >
          <TopTracks />
        </Provider>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("renders without crashing", () => {
      const div = document.createElement("div");
      ReactDOM.render(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopTracks: () => {
                  return Promise.reject({
                    body: {
                      items: [
                        { images: null, name: "No Image" },
                        { images: [{ url: "fakeUrl" }], name: "Image" }
                      ]
                    }
                  });
                }
              }
            }
          })}
        >
          <TopTracks />
        </Provider>,
        div
      );
      ReactDOM.unmountComponentAtNode(div);
    });
  });

  /**
   * Tests for the component with an error from Spotify.
   */
  describe("error from Spotify api", () => {
    it("matches the snapshot", () => {
      const component = renderer.create(
        <Provider
          store={configureStore({
            api: {
              spotifyApi: {
                getMyTopTracks: () => {
                  return Promise.reject("error");
                }
              }
            }
          })}
        >
          <TopTracks />
        </Provider>
      );
      let tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider
        store={configureStore({
          api: {
            spotifyApi: {
              getMyTopTracks: () => {
                return Promise.reject("error");
              }
            }
          }
        })}
      >
        <TopTracks />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
});
