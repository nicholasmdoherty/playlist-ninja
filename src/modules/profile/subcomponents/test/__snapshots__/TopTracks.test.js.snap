// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopTracks.jsx error from Spotify api matches the snapshot 1`] = `
<div
  className="p-3 row"
>
  <div
    className="text-center col-12"
  >
    <h3
      className="display-5"
    >
       Top Tracks 
    </h3>
    <div
      style={
        Object {
          "height": "4vh",
        }
      }
    />
  </div>
</div>
`;

exports[`TopTracks.jsx recieved top track data matches the snapshot 1`] = `
<div
  className="p-3 row"
>
  <div
    className="text-center col-12"
  >
    <h3
      className="display-5"
    >
       Top Tracks 
    </h3>
    <div
      style={
        Object {
          "height": "4vh",
        }
      }
    />
  </div>
</div>
`;

exports[`TopTracks.jsx with empty track array matches the snapshot 1`] = `
<div
  className="p-3 row"
>
  <div
    className="text-center col-12"
  >
    <h3
      className="display-5"
    >
       Top Tracks 
    </h3>
    <div
      style={
        Object {
          "height": "4vh",
        }
      }
    />
  </div>
</div>
`;
